### PR TITLE
[NGRM] - New layer (TriangleLayer) to represent unstructured surfaces #1429

### PR DESCRIPTION
**Pull request by nilscb**

### CODE REVIEW

File: react/src/lib/components/SubsurfaceViewer/layers/map/mapLayer.ts [hunk 1]
```diff
@@ -342,7 +342,7 @@ export default class MapLayer extends CompositeLayer<MapLayerProps<unknown>> {
                     const zMin = -meshZValueRange[0];
                     const xMax = xMin + xinc * xcount;
                     const yMax = yMin + yinc * ycount;
-                    const zMax = -meshZValueRange[0];
+                    const zMax = -meshZValueRange[1];
```
Comment (hkfb) [inline react/src/lib/components/SubsurfaceViewer/layers/map/mapLayer.ts:345]: Can you explain this change? Is there a bug in the map layer?

File: react/src/lib/components/SubsurfaceViewer/layers/triangle/webworker.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            normal_mean[2] += normal[2];
+        }
+
+        normalize(normal_mean as Vec);
+
+        vertexsNormals[3 * vertex_no + 0] = normal_mean[0];
+        vertexsNormals[3 * vertex_no + 1] = normal_mean[1];
+        vertexsNormals[3 * vertex_no + 2] = normal_mean[2];
+
+        // KEEP!
```
Comment (hkfb) [inline react/src/lib/components/SubsurfaceViewer/layers/triangle/webworker.ts:122]: Do you think you will need this again?

File: react/src/lib/components/SubsurfaceViewer/layers/triangle/webworker.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        drawMode: 1, // corresponds to GL.LINES,
+        attributes: {
+            positions: { value: new Float32Array(line_positions), size: 3 },
+        },
+        vertexCount: line_positions.length / 3,
+    };
+
+    //const t1 = performance.now();
+    // Keep this.
+    //console.log(`Task makeMesh took ${(t1 - t0) * 0.001}  seconds.`);
```
Comment (hkfb) [inline react/src/lib/components/SubsurfaceViewer/layers/triangle/webworker.ts:159]: Does this instrumentation reduce performance? If not we could consider not disabling it and use the `console.debug()` statement.

File: react/src/lib/components/SubsurfaceViewer/layers/triangle/webworker.ts (hunk 2 above)
Comment (nilscb) [reply react/src/lib/components/SubsurfaceViewer/layers/triangle/webworker.ts:122]: I was hoping to keep it for a while at least. It is a way to verify that the user has oriented all the triangles the right way. For now this layer is only tested with 4 triangles.

File: react/src/lib/components/SubsurfaceViewer/layers/triangle/webworker.ts (hunk 3 above)
Comment (nilscb) [reply react/src/lib/components/SubsurfaceViewer/layers/triangle/webworker.ts]: Fixed

File: react/src/lib/components/SubsurfaceViewer/layers/map/mapLayer.ts (hunk 1 above)
Comment (nilscb) [reply react/src/lib/components/SubsurfaceViewer/layers/map/mapLayer.ts:345]: Yes! I sneaked the fix in!

File: react/src/lib/components/SubsurfaceViewer/layers/triangle/privateTriangleLayer.ts [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    contours: [-1, -1],
+    isContoursDepth: true,
+    gridLines: false,
+    color: [100, 100, 255],
+    coordinateSystem: COORDINATE_SYSTEM.CARTESIAN,
+    depthTest: true,
+};
+
+// This is a private layer used only by the composite TriangleLayer
+export default class privateTriangleLayer extends Layer<
```
Comment (hkfb) [inline react/src/lib/components/SubsurfaceViewer/layers/triangle/privateTriangleLayer.ts:75]: Would suggest sticking to variable case convention, so PascalCase for classes and camelCase for variables

File: react/src/lib/components/SubsurfaceViewer/layers/triangle/triangleLayer.ts [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    depthTest: boolean;
+
+    /**  If true means that input z values are interpreted as depths.
+     * For example depth of z = 1000 corresponds to -1000 on the z axis. Default true.
+     */
+    ZIncreasingDownwards: boolean;
+
+    // Non public properties:
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    setReportedBoundingBox?: any;
```
Comment (hkfb) [inline react/src/lib/components/SubsurfaceViewer/layers/triangle/triangleLayer.ts:122]: Would be good to have the proper type.

File: react/src/lib/components/SubsurfaceViewer/layers/triangle/triangleLayer.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    docs: {
+        inlineStories: false,
+        iframeHeight: 500,
+    },
+};
+
+const north_arrow_layer = {
+    "@@type": "NorthArrow3DLayer",
+    id: "north-arrow-layer",
+};
```
Comment (hkfb) [inline react/src/lib/components/SubsurfaceViewer/layers/triangle/triangleLayer.stories.tsx:20]: Please stick to camelCase convention

File: react/src/lib/components/SubsurfaceViewer/layers/triangle/triangleLayer.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+import { isEqual } from "lodash";
+import { makeFullMesh } from "./webworker";
+
+export type Params = {
+    vertexArray: Float32Array;
+    indexArray: Uint32Array;
+    smoothShading: boolean;
+};
+
+async function load_mesh_and_properties(
```
Comment (hkfb) [inline react/src/lib/components/SubsurfaceViewer/layers/triangle/triangleLayer.ts:13]: Please stick to case convention

**Review by hkfb — APPROVED** [review]

**Pull request merged**